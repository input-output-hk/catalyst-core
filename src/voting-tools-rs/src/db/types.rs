//! Extra types defined in postgres
//!
//! db-sync defines a few types in postgres. When diesel generates a schema, it uses these types,
//! but since they don't exist on the Rust side, type checking fails. So we provide them in this
//! module
//!
//! Many types here have slightly wonky capitalization, but that's just what diesel expects

use diesel_derive_enum::DbEnum;

#[derive(Debug, Clone, Copy, DbEnum, SqlType)]
pub enum Rewardtype {
    Leader,
    Member,
    Reserves,
    Treasury,
    Refund,
}

#[derive(Debug, Clone, Copy, DbEnum, SqlType)]
pub enum Scriptpurposetype {
    Spend,
    Mint,
    Cert,
    Reward,
}

#[derive(Debug, Clone, Copy, DbEnum, SqlType)]
pub enum Scripttype {
    Multisig,
    Timelock,
    PlutusV1,
    PlutusV2,
}

#[derive(Debug, Clone, Copy, DbEnum, SqlType)]
pub enum Syncstatetype {
    Lagging,
    Following,
}
