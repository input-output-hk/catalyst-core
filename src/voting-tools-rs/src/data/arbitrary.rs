use crate::{
    validation::{cbor::cbor_to_bytes, hash::hash},
    Sig, Signature, VotingKey, VotingPurpose,
};

use super::{
    Nonce, PubKey, Registration, RewardsAddress, SignedRegistration, StakeKeyHex, TxId,
    VotingKeyHex,
};
use cardano_serialization_lib::chain_crypto::{AsymmetricKey, Ed25519, SigningAlgorithm};
use proptest::{arbitrary::StrategyFor, prelude::*, strategy::Map};

type Inputs = (Nonce, RewardsAddress, VotingKeyHex, [u8; 32], TxId);

impl Arbitrary for SignedRegistration {
    type Parameters = ();
    type Strategy = Map<StrategyFor<Inputs>, fn(Inputs) -> Self>;

    fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
        any::<Inputs>().prop_map(
            |(nonce, rewards_address, voting_key, stake_secret, tx_id)| {
                arbitrary_signed_registration(
                    nonce,
                    rewards_address,
                    voting_key,
                    stake_secret,
                    tx_id,
                )
            },
        )
    }
}

pub(crate) fn arbitrary_signed_registration(
    nonce: Nonce,
    rewards_address: RewardsAddress,
    voting_key: VotingKeyHex,
    stake_secret: [u8; 32],
    tx_id: TxId,
) -> SignedRegistration {
    let secret = Ed25519::secret_from_binary(&stake_secret).unwrap();
    let public = Ed25519::compute_public(&secret);

    let stake_key: [u8; 32] = public.as_ref().try_into().unwrap();

    let registration = Registration {
        voting_key: VotingKey::Direct(voting_key),
        stake_key: StakeKeyHex(PubKey(stake_key.to_vec())),
        rewards_address,
        nonce,
        voting_purpose: Some(VotingPurpose::CATALYST),
    };

    let reg_cbor = registration.to_cbor();
    let reg_bytes = cbor_to_bytes(&reg_cbor);
    let reg_hash = hash(&reg_bytes);

    let sig = Ed25519::sign(&secret, &reg_hash);
    let sig_bytes: [u8; 64] = sig.as_ref().try_into().unwrap();

    let signature = Signature {
        inner: Sig(sig_bytes),
    };

    SignedRegistration {
        registration,
        signature,
        stake_key_hash: vec![0; 29],
        tx_id,
        slot: 12345,
    }
}

impl Arbitrary for RewardsAddress {
    type Parameters = ();
    type Strategy = Map<StrategyFor<Vec<u8>>, fn(Vec<u8>) -> Self>;

    fn arbitrary_with((): Self::Parameters) -> Self::Strategy {
        any::<Vec<u8>>().prop_map(|vec| RewardsAddress(vec.into()))
    }
}
