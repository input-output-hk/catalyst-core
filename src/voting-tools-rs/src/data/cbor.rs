use ciborium::{cbor, value::Value};
use itertools::Itertools;

use super::{Nonce, Registration, RewardsAddress, StakeKeyHex, VotingKey, VotingPurpose};

impl Registration {
    pub(crate) fn to_cbor(&self) -> Value {
        let Self {
            voting_key,
            stake_key,
            rewards_address,
            nonce,
            voting_purpose,
        } = self;
        match voting_purpose {
            None => cbor!({
                61284 => {
                    1 => voting_key.to_cbor(),
                    2 => stake_key.clone().to_cbor(),
                    3 => rewards_address.to_cbor(),
                    4 => nonce.to_cbor(),
                }
            }),
            Some(voting_purpose) => cbor!({

                61284 => {
                    1 => voting_key.to_cbor(),
                    2 => stake_key.clone().to_cbor(),
                    3 => rewards_address.to_cbor(),
                    4 => nonce.to_cbor(),
                    5 => voting_purpose.to_cbor(),
                }
            }),
        }
        .unwrap()
    }
}

impl VotingKey {
    pub(crate) fn to_cbor(&self) -> Value {
        match self {
            Self::Direct(key) => cbor!(key.0).unwrap(),
            Self::Delegated(map) => {
                let vec = map.iter().collect_vec();
                cbor!(vec).unwrap()
            }
        }
    }
}

impl StakeKeyHex {
    pub(crate) fn to_cbor(&self) -> Value {
        cbor!(self.0 .0).unwrap()
    }
}

impl RewardsAddress {
    pub(crate) fn to_cbor(&self) -> Value {
        cbor!(self.0).unwrap()
    }
}

impl Nonce {
    pub(crate) fn to_cbor(self) -> Value {
        cbor!(self.0).unwrap()
    }
}

impl VotingPurpose {
    pub(crate) fn to_cbor(self) -> Value {
        cbor!(self.0).unwrap()
    }
}
