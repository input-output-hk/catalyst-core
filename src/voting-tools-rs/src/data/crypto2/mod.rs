mod pub_key;
pub use pub_key::PubKey;

mod sig;
pub use sig::Sig;
