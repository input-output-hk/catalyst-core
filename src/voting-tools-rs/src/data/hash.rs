use blake2::{digest::typenum::U32, Blake2b, Digest};

pub(crate) const BLAKE_2B_256_HASH_SIZE: usize = 256 / 8;

/// Helper function to generate the `blake2b_256` hash of a byte slice
pub(crate) fn hash(bytes: &[u8]) -> [u8; BLAKE_2B_256_HASH_SIZE] {
    type Blake2b256 = Blake2b<U32>;

    let mut hasher = Blake2b256::new();
    hasher.update(bytes);
    let result = hasher.finalize();
    result.into()
}

#[cfg(test)]
mod tests {
    use super::*;

    // these consts are taken from cip 15 test vectors
    const HASH_BYTES_HEX: &str = "a3d63f26cd94002443bc24f24b0a150f2c7996cd3a3fd247248de396faea6a5f";
    const METADATA_CBOR_HEX_BYTES: &str = "a119ef64a40158200036ef3e1f0d3f5989e2d155ea54bdb2a72c4c456ccb959af4c94868f473f5a002582086870efc99c453a873a16492ce87738ec79a0ebd064379a62e2c9cf4e119219e03581de0ae3a0a7aeda4aea522e74e4fe36759fca80789a613a58a4364f6ecef041904d2";

    fn hex_test(content_hex: &str, hash_hex: &str) {
        let content_bytes = hex::decode(content_hex).unwrap();
        let hash_bytes = hex::decode(hash_hex).unwrap();

        let hashed = hash(&content_bytes);

        assert_eq!(hashed.to_vec(), hash_bytes);
    }

    // Some test cases generated by cyber chef
    #[test]
    fn hashes_correctly() {
        hex_test(
            "0000",
            "9ee6dfb61a2fb903df487c401663825643bb825d41695e63df8af6162ab145a6",
        );
        hex_test(METADATA_CBOR_HEX_BYTES, HASH_BYTES_HEX);
    }
}
