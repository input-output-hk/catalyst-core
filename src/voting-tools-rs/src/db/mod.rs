use crate::model::{DbHost, DbName, DbPass, DbUser};

mod queries;
pub mod types;

// We need to allow this because custom type imports aren't always used in all tables

#[allow(unused_imports)]
#[cfg(test)]
pub mod schema;

#[allow(unused_imports)]
#[cfg(not(test))]
mod schema;

mod utils;

pub use inner::{Conn, Db};
use serde::Deserialize;

/// Information required to connect to a database
#[derive(Debug, Clone, Deserialize)]
pub struct DbConfig {
    /// The name of the database
    pub name: DbName,
    /// The user to connect with
    pub user: DbUser,
    /// The hostname to connect to
    pub host: DbHost,
    /// The corresponding password for this user
    pub password: Option<DbPass>,
}

/// Inner module to hide database internals from database code
mod inner {
    use super::DbConfig;
    use color_eyre::Result;
    use diesel::{
        pg::Pg,
        query_builder::QueryFragment,
        query_dsl::LoadQuery,
        r2d2::{ConnectionManager, Pool, PooledConnection},
        result::QueryResult,
        PgConnection,
    };
    use microtype::secrecy::Zeroize;

    /// Type alias for the connection type provided to diesel code
    pub type Conn = PooledConnection<ConnectionManager<PgConnection>>;

    /// A handle to the db-sync database instance
    pub struct Db(Pool<ConnectionManager<PgConnection>>);

    impl Db {
        /// Connect to the database with the provided credentials
        ///
        /// # Errors
        ///
        /// Returns an error if connecting to the database fails
        pub fn connect(
            DbConfig {
                name,
                user,
                host,
                password,
            }: DbConfig,
        ) -> Result<Self> {
            use microtype::secrecy::ExposeSecret;
            let mut password = password
                .map(|p| format!(":{}", p.expose_secret()))
                .unwrap_or_default();

            let url = format!("postgres://{user}{password}@{host}/{name}",);
            let manager = ConnectionManager::new(url);
            let pool = Pool::new(manager)?;

            password.zeroize();
            Ok(Db(pool))
        }

        /// Execute a query against the database
        pub(crate) fn exec<T, F>(&self, f: F) -> Result<T>
        where
            F: FnOnce(&mut Conn) -> QueryResult<T>,
        {
            let mut conn = self.0.get()?;
            let result = f(&mut conn)?;
            Ok(result)
        }
    }

    impl std::fmt::Debug for Db {
        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
            f.write_str("Db")
        }
    }

    /// Types which can be easily used as an ergonomic "query object"
    pub(super) trait DbQuery<'a, T>
    where
        Self: LoadQuery<'a, Conn, T> + QueryFragment<Pg> + Send + 'a,
    {
        fn sql_string(&self) -> String {
            let debug = diesel::debug_query(self);
            format!("{debug}")
        }
    }

    impl<'a, T, S> DbQuery<'a, T> for S where S: LoadQuery<'a, Conn, T> + QueryFragment<Pg> + Send + 'a {}
}
