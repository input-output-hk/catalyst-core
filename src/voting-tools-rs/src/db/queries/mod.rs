use crate::data::{SignedRegistration, SlotNo};
use crate::data_provider::DataProvider;
use crate::verify::StakeKeyHash;
use crate::Db;
use bigdecimal::BigDecimal;
use dashmap::DashMap;

mod stake_value;
pub(crate) mod staked_utxo_ada;
mod vote_registrations;

impl DataProvider for Db {
    fn vote_registrations(
        &self,
        lower: SlotNo,
        upper: SlotNo,
    ) -> color_eyre::Result<Vec<SignedRegistration>> {
        self.vote_registrations(lower, upper)
    }

    fn stake_values(&self, stake_addrs: &[StakeKeyHash]) -> DashMap<StakeKeyHash, BigDecimal> {
        self.stake_values(stake_addrs)
    }
}
