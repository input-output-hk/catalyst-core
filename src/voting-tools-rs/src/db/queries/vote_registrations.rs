use std::fs::File;
use std::io::{BufWriter, Write};

use crate::{db::inner::DbQuery, Db};

use crate::data::{SignedRegistration, SlotNo};
use crate::db::schema::{block, tx, tx_metadata};
use bigdecimal::{BigDecimal, FromPrimitive};
use color_eyre::eyre::{eyre, Result};
use color_eyre::{Help, Report};
use diesel::RunQueryDsl;
use diesel::{
    BoolExpressionMethods, ExpressionMethods, JoinOnDsl, PgAnyJsonExpressionMethods,
    PgJsonbExpressionMethods, QueryDsl,
};
use itertools::Itertools;
use once_cell::sync::Lazy;
use serde_json::Value;

use tracing::info;

static METADATA_KEY: Lazy<BigDecimal> = Lazy::new(|| BigDecimal::from_isize(61284).unwrap());
static SIGNATURE_KEY: Lazy<BigDecimal> = Lazy::new(|| BigDecimal::from_isize(61285).unwrap());

type Row = (i64, Option<Value>, Option<Value>, Option<i64>);

impl Db {
    /// This query is detailed in ``src/db/design/vote_registration_query.md``
    ///
    /// 'lower' is an optional inclusive lower bound. If `None`, `0` is used instead.
    /// 'upper' is an optional inclusive upper bound. If `None`, `i64::MAX` is used instead.
    ///
    /// Returns an error if either `lower` or `upper` is greater than `i64::MAX`
    ///
    /// # Errors
    ///
    /// Returns an error if either of `lower` or `upper` doesn't fit in an `i64`, or if a
    /// database-specific error occurs
    pub fn vote_registrations(
        &self,
        lower: SlotNo,
        upper: SlotNo,
    ) -> Result<Vec<SignedRegistration>> {
        let lower = lower.into_i64().ok_or_else(|| eyre!("invalid i64"))?;
        let upper = upper.into_i64().ok_or_else(|| eyre!("invalid i64"))?;
        let q = query(lower, upper);

        info!("Querying Registrations from DB from Slot# {lower} to {upper} Inclusive.");

        let rows = self.exec(move |conn| q.load(conn))?;

        let (oks, errs): (_, Vec<_>) = rows.into_iter().map(convert_row).partition_result();

        if errs.is_empty() {
            info!("No registration errors detected.");
        } else {
            let file = File::options()
                .write(true)
                .create(true)
                .truncate(true)
                .open("convert_row.err")
                .unwrap();
            let mut writer = BufWriter::new(file);
            warn!("writing {} errors to convert_row.err", errs.len());
            for err in errs {
                writeln!(&mut writer, "==========================").unwrap();
                writeln!(&mut writer, "{err:#?}").unwrap();
            }
        }

        Ok(oks)
    }
}

/// This query doesn't exactly match the query in the doc. In particular, some filtering of invalid
/// JSONs isn't performed, since it breaks something (not sure what, just get no rows back). For
/// now, we'll just do the filtering in Rust
fn query(lower: i64, upper: i64) -> impl DbQuery<'static, Row> {
    let (meta_table, sig_table) = alias!(tx_metadata as meta_table, tx_metadata as sig_table);

    let metadata = meta_table.field(tx_metadata::json);
    let signature = sig_table.field(tx_metadata::json);

    let tables = meta_table
        .inner_join(tx::table.on(tx::id.eq(meta_table.field(tx_metadata::tx_id))))
        .inner_join(
            sig_table.on(sig_table
                .field(tx_metadata::tx_id)
                .eq(meta_table.field(tx_metadata::tx_id))),
        )
        .inner_join(block::table.on(block::id.eq(tx::block_id)));

    let selection = (
        meta_table.field(tx_metadata::tx_id),
        metadata,
        signature,
        block::slot_no,
    );

    let signature_keys_predicate = signature.has_all_keys(["1"].as_slice());
    let block_number_predicate = block::slot_no.ge(lower).and(block::slot_no.le(upper));

    let metadata_2 = metadata.retrieve_as_object("2");

    tables
        .filter(meta_table.field(tx_metadata::key).eq(&*METADATA_KEY))
        .filter(sig_table.field(tx_metadata::key).eq(&*SIGNATURE_KEY))
        .filter(signature_keys_predicate)
        .filter(block_number_predicate)
        .select(selection)
        .distinct_on(metadata_2)
}

/// Attempt to parse a row into a [`SignedRegistration`] struct
fn convert_row((tx_id, metadata, signature, _slot_no): Row) -> Result<SignedRegistration, Report> {
    let metadata = metadata.ok_or_else(|| eyre!("no metadata"))?;
    let signature = signature.ok_or_else(|| eyre!("no metadata"))?;

    let tx_id = u64::try_from(tx_id)?.into();
    let registration = serde_json::from_value(metadata.clone()).map_err(|e| {
        Report::from(e).with_warning(|| format!("registration json: {metadata:#?}"))
    })?;
    let signature = serde_json::from_value(signature.clone())
        .map_err(|e| Report::from(e).with_warning(|| format!("signature json: {signature:#?}")))?;

    Ok(SignedRegistration {
        registration,
        signature,
        stake_key_hash: vec![0; 29],
        tx_id,
        slot: 0,
    })
}

#[cfg(test)]
mod tests {
    use serde_json::{from_value, json};

    use crate::{
        data::{PubKey, Registration},
        VotingKey,
    };

    use super::*;

    fn cip15_test_vector_meta() -> serde_json::Value {
        json!({
            "1": "0036ef3e1f0d3f5989e2d155ea54bdb2a72c4c456ccb959af4c94868f473f5a0",
            "2": "86870efc99c453a873a16492ce87738ec79a0ebd064379a62e2c9cf4e119219e",
            "3": "e0ae3a0a7aeda4aea522e74e4fe36759fca80789a613a58a4364f6ecef",
            "4": 1234,
        })
    }

    fn cip15_test_vector_sig() -> serde_json::Value {
        json!({
            "1": "6c2312cd49067ecf0920df7e067199c55b3faef4ec0bce1bd2cfb99793972478c45876af2bc271ac759c5ce40ace5a398b9fdb0e359f3c333fe856648804780e",
        })
    }

    // #[test]
    // fn process_accepts_cip15_test_vector() {
    //     let rows = vec![(
    //         1,
    //         Some(cip15_test_vector_meta()),
    //         Some(cip15_test_vector_sig()),
    //         None,
    //     )];
    //
    //     let regs = vec![SignedRegistration {
    //         tx_id: 1.into(),
    //         registration: from_value(cip15_test_vector_meta()).unwrap(),
    //         signature: from_value(cip15_test_vector_sig()).unwrap(),
    //     }];
    //
    //     let result: Vec<_> = rows.into_iter().filter_map(convert_row).collect();
    //
    //     assert_eq!(result, regs);
    // }

    #[test]
    fn filters_bad_rows() {
        fn check(row: Row) {
            assert!(convert_row(row).is_err());
        }

        // bad sig
        check((
            1,
            Some(cip15_test_vector_meta()),
            Some(json!("random json")),
            None,
        ));

        // bad meta
        check((
            1,
            Some(json!("random json")),
            Some(cip15_test_vector_sig()),
            None,
        ));

        // none
        check((1, None, None, None));
    }

    #[test]
    fn query_snapshot() {
        let q = query(123, 234);
        insta::assert_snapshot!(q.sql_string());
    }

    #[test]
    fn can_parse_voting_key() {
        color_eyre::install().unwrap();
        let vps = json!([[
            "0x1b92110d6c2aee00b7b208992efe3a2511f6811577155dd89977e8600fa5c27a",
            1
        ]]);
        let _: VotingKey = from_value(vps).unwrap();
        let x = "0x0cf8d92c3fb4474942554a3c97f9a0421b69073fed4a600a6a77a1d72a8a";
        hex::decode(x.trim_start_matches("0x")).unwrap();

        let _: PubKey = from_value(json!(
            "0x0cf8d92c3fb4474942554a3c97f9a0421b69073fed4a600a6a77a1d72a8a"
        ))
        .unwrap();

        let value = json!({
            "1": [[
                "0x0cf8d92c3fb4474942554a3c97f9a0421b69073fed4a600a6a77a1d72a8a",
                1
            ]],
            "2": "0x19649e69c2fc0942d41452ad771278e0da18d1965bc9cc00d92d0e1a0f632ea4",
            "3": "0x005e868b277b468dcdec76f410fb9159e61565b3d61800dcfaab08afb7c17cbc9ffe7a69542d440e12fe800c708530e5a36c5f148c2e795f79",
            "4": 15_159_614,
            "5": 0,
        });
        let _: Registration = from_value(value).unwrap();
    }
}
