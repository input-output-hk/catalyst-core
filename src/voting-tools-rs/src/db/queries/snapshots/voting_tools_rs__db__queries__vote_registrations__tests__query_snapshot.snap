---
source: src/voting-tools-rs/src/db/queries/vote_registrations.rs
expression: q.sql_string()
---
SELECT DISTINCT ON (("meta_table"."json" -> $1))"meta_table"."tx_id", "meta_table"."json", "sig_table"."json", "block"."slot_no" FROM ((("tx_metadata" AS "meta_table" INNER JOIN "tx" ON ("tx"."id" = "meta_table"."tx_id")) INNER JOIN "tx_metadata" AS "sig_table" ON ("sig_table"."tx_id" = "meta_table"."tx_id")) INNER JOIN "block" ON ("block"."id" = "tx"."block_id")) WHERE (((("meta_table"."key" = $2) AND ("sig_table"."key" = $3)) AND ("sig_table"."json" ?& $4)) AND (("block"."slot_no" >= $5) AND ("block"."slot_no" <= $6))) -- binds: ["2", BigDecimal("61284"), BigDecimal("61285"), ["1"], 123, 234]
